dialogue-core/src/main/java/com/palantir/dialogue/core/Config.java

```diff
@@ -40,11 +42,17 @@
     default ClientConfiguration clientConf() {
         return ClientConfiguration.builder()
                 .from(rawConfig())
+                .uris(rawConfig().uris().stream().map(MeshMode::stripMeshPrefix).collect(Collectors.toList()))
                 .taggedMetricRegistry(
                         VersionedTaggedMetricRegistry.create(rawConfig().taggedMetricRegistry()))
                 .build();
     }
 
+    @Value.Derived
+    default MeshMode mesh() {
```

Comment:
can we keep all of the meshmode stuff internal to DialogueChannel? i.e. perform the computation internally instead of having a derived field on the config object?

Reply:
Sure we could, but then there's a possibility of constructing a Config object which is in a non-sensical state - i.e. has `mesh-` prefixed uris, but _doesn't_ have `MeshMode.USE_EXTERNAL_MESH` set.  The reason I think it's appropriate & desirable here is to emphasise to people that this isn't an extra degree of freedom in how you build a Config object, instead it's just purely _derived_ from the URIs that you already put in.

Otherwise, to keep this safety, we'd just have to put in a Value.Check that computes the mesh mode.

Third option is to not even have the mesh() param on this object, which would then require plumbing it in everywhere that we currently pass `Config cf`.

Reply:
At least within the scope of this PR it seems like there are only 2 places that actually care about the mesh mode stuff, and in both cases they didn't previous require the entire `Config` object to be initialized. So it doesn't seem like we're saving much by pushing the mesh-mode stuff into the config object and it just exposes additional API on the `Config`.

Reply:
Config is package private, so the only people who will be affected by this decision will be us / future maintainers of dialogue.

Back [in april](https://github.com/palantir/dialogue/pull/690) I found dialogue channel had grown to become really hard to read, so condensed all of the 'inputs' into this one Config object. That enables the pattern where Channels are created with this convention:

```
HostMetricsChannel.create(cf, channel, uri)
QueuedChannel.create(cf, endpoint, limited)
DeprecationWarningChannel.create(cf, channel, endpoint);
TimingEndpointChannel.create(cf, channel, endpoint);
```

In each of these cases, `Config cf` is only plumbed to the static factory method, so the actual Channel's construct has only the bits it really needs (e.g. a clock or a channelname string etc), which makes it convenient to supply only exactly what you need for unit-testing.

Essentially I think this convention helps us minimize plumbing, which keeps the top level DialogueChannel as concise as possible, so that you can follow which channels are wired into which other channels, without getting distracted.